Added Reindex and deprecated CrossIndex

giovannialbero1992 opened the pull request:
[#1311](https://github.com/ruflin/Elastica/issues/1311)

ruflin commented on lib/Elastica/Reindex.php:
```diff
@@ -0,0 +1,117 @@
+    {
+        $body = self::getBody($oldIndex, $newIndex, $options);
+
+        $reindexEndpoint = new \Elasticsearch\Endpoints\Reindex();
```
glad to see you used the endpoints 👍

ruflin commented on lib/Elastica/Reindex.php:
```diff
@@ -0,0 +1,117 @@
+    const SIZE = 'size';
+    const QUERY = 'query';
+
+    public static function reindex(Index $oldIndex, Index $newIndex, array $options = [])
```
Why did you use static method here? I'm thinking the implementation would become simpler if you would use a Reindex object where old, new and options are passed.

giovannialbero1992 replied:
I've used static method because there are not external depenendencies but maybe, to be compliant with OOP I should remove static. Do you suggest to pass parameters into constructor or in reindex method ?

ruflin replied:
I was thinking passing params to constructor and then having something like `$reindex->Run()` or `Start` or something similar. Perhaps Start() could also take additional options. Not 100% sure yet how the most common use case will look like. Any thoughts?

giovannialbero1992 replied:
I think that call `new` and `->run()` (or `->start()`) are ever near to constructor and there is no need to change options after have called `new Reindex(...)`. Maybe in a first time we can see how community uses reindex and if there will be need to change the use, we will change it consequently.

giovannialbero1992 commented:
@ruflin I've removed static from class and renamed method in `run`. What do you think about this version?

ruflin commented on lib/Elastica/Reindex.php:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * @var Index
+     */
+    private $oldIndex;
```
We always use $_oldIndex for all the variable names which are private / protected but perhaps the time has come to break this as not much other projects are left with the old pattern.

giovannialbero1992 replied:
I changed to be compliant with old pattern... maybe we can change the pattern into another commit on all protected/private variables :)

ruflin commented on lib/Elastica/Reindex.php:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * @var Index
+     */
+    private $newIndex;
```
Please make these all protected and not private. In case someone wants to extend Reindex in his code, he should be able to.

ruflin commented on lib/Elastica/Reindex.php:
```diff
@@ -0,0 +1,139 @@
+        return $this->newIndex;
+    }
+
+    private function getBody($oldIndex, $newIndex, $options)
```
Make these also protected, not private.

giovannialbero1992 commented:
@ruflin, I've changed it :)

ruflin commented on lib/Elastica/Reindex.php:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * @var array
+     */
+    private $_options;
```
This one should also be protected.

ruflin commented on lib/Elastica/Reindex.php:
```diff
@@ -0,0 +1,139 @@
+        return $this->_newIndex;
+    }
+
+    protected function getBody($oldIndex, $newIndex, $options)
```
This would be _getBody(

ruflin commented on lib/Elastica/Reindex.php:
```diff
@@ -0,0 +1,139 @@
+        ], $this->resolveDestOptions($options));
+    }
+
+    private static function resolveSourceOptions(array $options)
```
This should also be not static and start with _* Also make it protected if there is no need to make it private.

giovannialbero1992 commented:
@ruflin I've changed all names with '_' and removed a missing static from method.
`_getDestPartBody`, `_getSourcePartBody` and `_getBody` now are protected. Thanks for your reviews :)

ruflin commented:
@giovannialbero1992 Thanks for your contribution.
